### ISSUE TITLE
Percent-encode paths

### ISSUE THREAD

**Issue by binEpilo** [jannis-baum/vivify.vim#20]
### Description

I just added this plugin to my plugins, installed vivify on my arch system (viv --version: vivify-server v0.6.0-1-aur) with yay and with :Vivify firefox opens my file correctly. Still there is one problem: for some reason when I now edit the file with neovim those changes arent automatically shown in the vivify firefox preview: I have to save the file with :w and then reload the page with Ctrl+R. I know that this is definitely user error on my side, but I really dont know what I need to configure here. (I also tried adding the line vim.g.vivify_instant_refresh = 1 to my init.lua)

**Comment by binEpilo** [comment]
With this config it works now. So this is not an issue with vivify, will try to fix the issue now for my config :)

**Comment by tuurep** [comment]
Great, thanks :smile:

I am kinda interested in what went wrong so feel free to share if you find out. Or if I can help with something.

**Comment by binEpilo** [comment]
I think the issue might be that I am using file names with spaces in them with vivify. While I can open them with :Vivify and they appear in the browsers, vivify doesnt update/refresh them when typing for some reason

**Comment by tuurep** [comment]
Yes, I can reproduce this! Nice find.

@jannis-baum did we have some fix for this that can be used in this plugin's side as well?

**Comment by tuurep** [comment]
Relevant: https://github.com/jannis-baum/Vivify/issues/26

**Comment by jannis-baum** [comment]
Ah, yes, this plugin is not doing the percent-encoding. Have to adjust

https://github.com/jannis-baum/vivify.vim/blob/main/autoload/vivify.vim#L30

and percent-encode the file path. I haven't looked into how to do that with Vimscript.

**Comment by jannis-baum** [comment]
I did a quick search, looks like Vimscript doesn't offer a native way of doing this. So from what I can tell the main options would be:

1. implement the percent-encoding ourselves in Vimscript, e.g. [like this](https://vi.stackexchange.com/questions/42717/how-do-i-turn-a-string-into-url-encoded-string) (this example might be a bit more lengthy than it would have to be)
2. require that Vim is compiled with Python[^1] for this plugin and use Python's native `urllib` to do it as a one-liner [like this](https://stackoverflow.com/a/9345102) in Vimscript's `py3eval` function
3. run `sed` or `awk` as a subprocess to do it [like this](https://askubuntu.com/a/295312)

1 and 3 have wouldn't introduce any new dependency but are slow, 2 is the cleanest but does introduce the Python dependency (although AFAIK it's quite uncommon to have a version of Vim without Python). 1 has the disadvantage of having to implement this ourselves, while 3 has the disadvantage of having to call a subprocess. My preference would be 2. @tuurep what do you think?

[^1]: You can use `vim --version` and check that it says `+python3` somewhere to verify your Vim was compiled with Python3

**Comment by tuurep** [comment]
Ah, I see

Sounds good to use the python oneliner, apparently it works on neovim too,

But this makes me think about whether I should make a separate `vivify.nvim` in Lua, in case Lua has a clean way to do the url encoding.

As a bonus, we'd get rid of the ambiguity with for example this vim/nvim difference:

https://github.com/jannis-baum/vivify.vim/blob/822b527435781a82b47f056ae7edee83689e763c/autoload/vivify.vim#L3-L4

As a more long term idea, do you think I should go for it?

**Comment by jannis-baum** [comment]
I checked and it seemed like Lua doesn't have anything, I only found manual/custom implementations for it.

Anyways, I like that we can maintain and test this plugin together. If we separate Vim and nvim I think I'd be by myself here and also couldn't easily help you with the other one anymore haha. So I'd prefer to keep it as one ^^

**Comment by tuurep** [comment]
> only found manual/custom implementations for it.

Yeah, in that case, lets just keep it as one :D

I don't mind installing Vim to test Vim-specific stuff however. But yeah, lets do it that way.

**Comment by tuurep** [comment]
Hey @binEpilo sorry that this took longer than expected (at least I intended to do it sooner :smile: )

We've solved the spaces-in-filenames issue in #21, feel free to test as well or give feedback!

The way it's done is that if you have python3 support in (n)vim, we do a proper encoding using a library

But if you don't have python3 support that's fine too, in that case we just substitute spaces to `%20` for the url. The only downside we ran into with not "encoding properly" was that if we had a file with curly braces in the filename, those wouldn't work.

But enabling python3 will handle cases like that too.

Meaning:

- Vim needs to be compiled with `+python`
- NeoVim needs [pynvim](https://github.com/neovim/pynvim) installed
  - (for me on Arch Linux, I needed to install package `python-pynvim` from pacman repos)
